cf-execd/cf-execd-runagent.c

```diff
@@ -0,0 +1,150 @@
+
+  This file is part of CFEngine 3 - written and maintained by Northern.tech AS.
+
+  This program is free software; you can redistribute it and/or modify it
+  under the terms of the GNU General Public License as published by the
+  Free Software Foundation; version 3.
+
+  This program is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with this program; if not, write to the Free Software
+  Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA
+
+  To the extent this program is licensed as part of the Enterprise
+  versions of CFEngine, the applicable Commercial Open Source License
+  (COSL) may apply to this file if you as a licensee so wish it. See
+  included file COSL.txt.
+*/
+
+#include <stdio.h>
+
+#include <logging.h>
+#include <buffer.h>
+#include <json.h>
+#include <pipes.h>
+#include <alloc.h>
+#include <file_lib.h>
+#include <known_dirs.h>
+
+#include <cf-execd-runagent.h>
+
+
+/**
+ * Handle request to run cf-runagent.
+ *
+ * @note Expected to be called from a forked child process (blocks and forks).
+ */
+void HandleRunagentRequest(int conn_fd)
+{
+    FILE *conn_file = fdopen(conn_fd, "r+");
+    if (conn_file == NULL)
+    {
+        Log(LOG_LEVEL_ERR, "Failed to create file stream from the connection file descriptor: %s",
+            GetErrorStr());
+
+        /* SIGH: There's no dprintf() on exotics. */
+        char error_buf[256];
+        int length = snprintf(error_buf, sizeof(error_buf),
+                              "{\"error\": \"Failed to create file stream from the connection file descriptor: %s\"}",
+                              GetErrorStr());
+        assert(length < sizeof(error_buf));
+        write(conn_fd, error_buf, MIN(sizeof(error_buf), (size_t) length));
+        close(conn_fd);
+        return;
+    }
+
+    char *request = NULL;
+    size_t line_size = 0;
+    ssize_t n_read = getline(&request, &line_size, conn_file);
+    if (n_read <= 0)
+    {
+        Log(LOG_LEVEL_ERR, "Failed to read the runagent request: %s",
+            (n_read < 0) ? GetErrorStr() : "no data");
+        fprintf(conn_file, "{\"error\": \"Failed to read the runagent request: %s\"}",
+                (n_read < 0) ? GetErrorStr() : "no data");
+        free(request);
+        fclose(conn_file);
+        return;
+    }
+
+    /* TODO: '--raw' for just getting output from cf-runagent directly? */
+    for (char *c = (request + n_read - 1); (c > request) && (*c != '\n'); c--)
+    {
+        if (*c == '\n')
+        {
+            *c = '\0';
+        }
+    }
+
+    char *command;
+    int ret = xasprintf(&command, "%s%ccf-runagent -b -H %s", GetBinDir(), FILE_SEPARATOR, request);
+    assert(ret != -1); /* should never happen */
+    if (ret == -1)
```

Comment:
I don't understand why we assert the return value from `xasprintf` with a comment saying "this should never happen", and then handle it as if it could happen? Does not `xasprintf` call `DoCleanupAndExit`, thus the error handling will never happen?

Reply:
Good point! I'll gladly drop this. :+1: :grin: